sigal/plugins/zip_gallery.py

```diff
@@ -0,0 +1,90 @@
+# all copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+
+""" This plugin controls the generation of a ZIP archive for a gallery
+
+To ignore a ZIP gallery generation for a particular album, put a ``.nozip_gallery`` file next to it in its parent
+folder. Only the existence of this ``.nozip_gallery`` file is tested.
+"""
+
+import logging
+import os
+from os.path import isfile, join, splitext
+from sigal import signals
+from sigal.gallery import Album
+from sigal.utils import cached_property
+import zipfile
+
+logger = logging.getLogger(__name__)
+
+def noop_zip_method(album):
+    """A zip method that does nothing at all"""
+    return False
+
+def generate_album_zip_method(album):
+    """Make a ZIP archive with all media files and return its path.
+
+    If the ``zip_gallery`` setting is set,it contains the location of a zip
+    archive with all original images of the corresponding directory.
+
+    """
+
+    zip_gallery = album.settings['zip_gallery']
+
+    if zip_gallery and len(album) > 0:
+        zip_gallery = zip_gallery.format(album=album)
+        archive_path = join(album.dst_path, zip_gallery)
+        if (album.settings.get('zip_skip_if_exists', False) and
+                isfile(archive_path)):
+            album.logger.debug("Archive %s already created, passing",
+                              archive_path)
+            return zip_gallery
+
+        archive = zipfile.ZipFile(archive_path, 'w', allowZip64=True)
+        attr = ('src_path' if album.settings['zip_media_format'] == 'orig'
+                else 'dst_path')
+
+        for p in album:
+            path = getattr(p, attr)
+            try:
+                archive.write(path, os.path.split(path)[1])
+            except OSError as e:
+                album.logger.warn('Failed to add %s to the ZIP: %s', p, e)
+
+        archive.close()
+        album.logger.debug('Created ZIP archive %s', archive_path)
+        return zip_gallery
+
+
+def nozip_galery_file(album, settings=None):
+    """Filesystem based switch to disable ZIP generation for an Album"""
+    nozipgallerypath = os.path.join(album.src_path, ".nozip_gallery")
+
+    zip_method = generate_album_zip_method
+
+    if os.path.isfile(nozipgallerypath):
+        logger.info("Ignoring ZIP gallery generation for album '%s' because of present "
+                    ".nozip_gallery file", album.name)
+        zip_method = noop_zip_method
+
+    album.__class__.zip = cached_property(zip_method)
```

Comment:
Since the property is set on the class, it will be used for all the Album objects (the last will win). So instead I think that you can check for the "nozip" file directly in `generate_album_zip_method` ?

Reply:
You're absolutely right, this was too naive. I've switched to a single _function_ that does all the checks and contains all the logic.